kart/s3_util.py

```diff
@@ -0,0 +1,106 @@
+from base64 import standard_b64decode
+import functools
+import os
+from pathlib import Path
+import tempfile
+from urllib.parse import urlparse
+
+import boto3
+import click
+
+from kart.exceptions import NotFound, NO_IMPORT_SOURCE
+
+# Utility functions for dealing with S3 - not yet launched.
+
+
+@functools.lru_cache(maxsize=1)
+def get_s3_config():
+    # TODO - add an option --s3-region to commands where it would be useful.
+    return None
+
+
+@functools.lru_cache(maxsize=1)
+def get_s3_client():
+    client = boto3.client("s3", config=get_s3_config())
+    if "AWS_NO_SIGN_REQUEST" in os.environ:
+        client._request_signer.sign = lambda *args, **kwargs: None
+    return client
+
+
+@functools.lru_cache(maxsize=1)
+def get_s3_resource():
+    resource = boto3.resource("s3", config=get_s3_config())
+    if "AWS_NO_SIGN_REQUEST" in os.environ:
+        resource.meta.client._request_signer.sign = lambda *args, **kwargs: None
+    return resource
+
+
+@functools.lru_cache()
+def get_bucket(name):
+    return get_s3_resource().Bucket(name)
+
+
+def fetch_from_s3(s3_url, output_path=None):
+    """
+    Downloads the object at s3_url to output_path.
+    If output-path is not set, creates a temporary file using tempfile.mkstemp()
+    """
+    # TODO: handle failure.
+    parsed = urlparse(s3_url)
+    bucket = get_bucket(parsed.netloc)
+    if output_path is None:
+        fd, path = tempfile.mkstemp()
+        # If we keep it open, boto3 won't be able to write to it (on Windows):
+        os.close(fd)
+        output_path = Path(path)
+    bucket.download_file(parsed.path.lstrip("/"), str(output_path.resolve()))
+    return output_path
+
+
+def expand_s3_glob(source_spec):
+    """
+    Given an s3_path with '*' wildcard in, uses prefix and suffix matching to find all S3 objects that match.
+    Subdirectories (or the S3 equivalent - S3 is not exactly a directory hierarchy) are not matched -
+    that is, s3://bucket/path/*.txt matches s3://bucket/path/example.txt but not s3://bucket/path/subpath/example.txt
+    """
+    # TODO: handle any kind of failure, sanity check to make sure we don't match a million objects.
```

Comment:
meh; a million objects is actually fine at this level, not really our job to try to guess if that's the user's intent 🤷